python/mspasspy/algorithms/basic.py

```diff
@@ -1,5 +1,55 @@
 from mspasspy.util.decorators import mspass_func_wrapper
 from mspasspy.ccore.algorithms.basic import LinearTaper, CosineTaper, VectorTaper
+from mspasspy.ccore.utility import MsPASSError, ErrorSeverity
+from mspasspy.util import logging_helper
+import mspasspy.ccore.algorithms.basic as bsc
+from mspasspy.ccore.seismic import (
+    TimeSeries,
+    TimeSeriesEnsemble,
+    Seismogram,
+    SeismogramEnsemble,
+)
+
+
+@mspass_func_wrapper
+def ExtractComponent(
+    data,
+    component,
+    *args,
+    object_history=False,
+    alg_name=None,
+    alg_id=None,
+    dryrun=False,
+    inplace_return=False,
+    function_return_key=None,
+    **kwargs
+):
+    if isinstance(data, Seismogram):
+        empty = TimeSeries()
+        empty.kill()
+        if data.dead():
+            return empty
+        try:
+            d = bsc.ExtractComponent(data, component)
+            return d
+        except Exception as err:
+            data.elog.log_error("ExtractComponent", str(err), ErrorSeverity.Invalid)
+            return empty
+    elif isinstance(data, SeismogramEnsemble):
```

Comment:
This has the same issue as above. Although we currently don't have a history for the ensemble type, the members of the ensemble may still container history that should be retained.

Reply:
Here I let each member in the TimeSeriesEnsemble load history of the corresponding member in the input SeismogramEnsemble. When the input is dead or an exception occurs, the return object will have elog inherited.